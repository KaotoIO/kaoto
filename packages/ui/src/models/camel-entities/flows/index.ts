export * from './route';
export * from './step';
