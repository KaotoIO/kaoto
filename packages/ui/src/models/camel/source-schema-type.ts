export enum SourceSchemaType {
  Route = 'Route',
  Integration = 'Integration',
  Kamelet = 'Kamelet',
  KameletBinding = 'KameletBinding',
  Pipe = 'Pipe',
}
