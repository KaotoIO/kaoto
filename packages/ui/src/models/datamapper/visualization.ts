import { IDocument, IField, PrimitiveDocument } from './document';
import { ExpressionItem, FieldItem, IFunctionDefinition, MappingItem, MappingParentType, MappingTree } from './mapping';
import { DocumentType, NodePath } from './path';
import { RefObject } from 'react';

export interface NodeData {
  title: string;
  id: string;
  path: NodePath;
  isSource: boolean;
  isPrimitive: boolean;
}

export interface TargetNodeData extends NodeData {
  mappingTree: MappingTree;
  mapping?: MappingParentType;
}

export type SourceNodeDataType = DocumentNodeData | FieldNodeData;
export type TargetNodeDataType = TargetDocumentNodeData | TargetFieldNodeData;

export class DocumentNodeData implements NodeData {
  constructor(document: IDocument) {
    this.title = document.documentId;
    this.id = `doc-${document.documentType}-${document.documentId}`;
    this.path = NodePath.fromDocument(document.documentType, document.documentId);
    this.document = document;
    this.isSource = document.documentType !== DocumentType.TARGET_BODY;
    this.isPrimitive = document instanceof PrimitiveDocument;
  }

  document: IDocument;
  title: string;
  id: string;
  path: NodePath;
  isSource: boolean;
  isPrimitive: boolean;
}

export class TargetDocumentNodeData extends DocumentNodeData implements TargetNodeData {
  constructor(document: IDocument, mappingTree: MappingTree) {
    super(document);
    this.mappingTree = mappingTree;
    this.mapping = mappingTree;
  }
  mappingTree: MappingTree;
  mapping: MappingTree;
}

export class FieldNodeData implements NodeData {
  constructor(
    public parent: NodeData,
    public field: IField,
  ) {
    this.title = field.name;
    this.id = field.id;
    this.path = NodePath.childOf(parent.path, this.id);
    this.isSource = parent.isSource;
    this.isPrimitive = parent.isPrimitive;
  }

  title: string;
  id: string;
  path: NodePath;
  isSource: boolean;
  isPrimitive: boolean;
}

export class TargetFieldNodeData extends FieldNodeData implements TargetNodeData {
  constructor(
    public parent: TargetNodeData,
    public field: IField,
    public mapping?: FieldItem,
  ) {
    super(parent, field);
    this.mappingTree = parent.mappingTree;
  }
  mappingTree: MappingTree;
}

export class MappingNodeData implements TargetNodeData {
  constructor(
    public parent: TargetNodeData,
    public mapping: MappingItem,
  ) {
    this.title = mapping.name;
    this.id = mapping.id;
    this.path = NodePath.childOf(parent.path, this.id);
    this.isPrimitive = parent.isPrimitive;
    this.mappingTree = parent.mappingTree;
  }

  title: string;
  id: string;
  path: NodePath;
  isSource = false;
  isPrimitive: boolean;
  mappingTree: MappingTree;
}

export class FieldItemNodeData extends MappingNodeData {
  constructor(
    public parent: TargetNodeData,
    public mapping: FieldItem,
  ) {
    super(parent, mapping);
    this.title = mapping.field.name;
    this.field = mapping.field;
  }
  public field: IField;
}

export class AddMappingNodeData implements TargetNodeData {
  constructor(
    public parent: TargetNodeData,
    public field: IField,
  ) {
    const ID_PREFIX = 'add-mapping-';
    this.id = ID_PREFIX + field.id;
    this.path = NodePath.childOf(parent.path, this.id);
    this.title = field.name;
    this.isPrimitive = parent.isPrimitive;
    this.mappingTree = parent.mappingTree;
  }
  id: string;
  isPrimitive: boolean;
  isSource = false;
  mappingTree: MappingTree;
  path: NodePath;
  title: string;
}

class SimpleNodePath extends NodePath {
  constructor(public path: string) {
    super();
  }
  toString() {
    return this.path;
  }
}
export class EditorNodeData implements NodeData {
  constructor(public mapping: ExpressionItem) {}
  id: string = 'editor';
  isPrimitive: boolean = false;
  isSource: boolean = false;
  path: NodePath = new SimpleNodePath('Editor');
  title: string = 'Editor';
}

export class FunctionNodeData implements NodeData {
  constructor(public functionDefinition: IFunctionDefinition) {
    this.id = functionDefinition.name;
    this.title = functionDefinition.displayName;
    this.path = new SimpleNodePath('Func:' + functionDefinition.name);
  }

  id: string;
  isPrimitive: boolean = false;
  isSource: boolean = true;
  path: NodePath;
  title: string;
}

export interface IMappingLink {
  sourceNodePath: string;
  targetNodePath: string;
  isSelected: boolean;
}

export interface NodeReference {
  path: string;
  isSource: boolean;
  headerRef: HTMLDivElement | null;
  containerRef: HTMLDivElement | null;
}

export type LineCoord = {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
};

export type LineProps = LineCoord & {
  sourceNodePath: string;
  targetNodePath: string;
  isSelected?: boolean;
  svgRef?: RefObject<SVGSVGElement>;
};
