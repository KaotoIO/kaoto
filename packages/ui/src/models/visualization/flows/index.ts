export * from './camel-catalog.service';
export * from './pipe-visual-entity';
export * from './kamelet-binding-visual-entity';
export * from './kamelet-visual-entity';
export * from './camel-route-visual-entity';
