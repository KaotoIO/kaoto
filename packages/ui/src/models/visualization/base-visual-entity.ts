import { DefinedComponent } from '../camel-catalog-index';
import { BaseCamelEntity, EntityType } from '../camel/entities';
import { KaotoSchemaDefinition } from '../kaoto-schema';

/**
 * BaseVisualCamelEntity
 *
 * This interface is used to represent a visual Camel entity.
 * f.i. Camel Route, Kamelet, KameletBinding, etc.
 * All dedicated Camel code should be implemented using this interface.
 */
export interface BaseVisualCamelEntity extends BaseCamelEntity {
  id: string;
  type: EntityType;

  getId: () => string;

  setId: (id: string) => void;

  /** Given a path, get the component label */
  getNodeLabel: (path?: string, labelType?: string) => string;

  /** Given a path, get the component tooltip content */
  getTooltipContent: (path?: string) => string;

  /** Given a path, get the component type and definition */
  getComponentSchema: (path?: string) => VisualComponentSchema | undefined;

  /** Returnt fields that should be omitted when configuring this entity */
  getOmitFormFields: () => string[];

  /** Given a path, update the model */
  updateModel(path: string | undefined, value: unknown): void;

  /** Add a step to the underlying Camel entity */
  addStep: (options: {
    definedComponent: DefinedComponent;
    mode: AddStepMode;
    data: IVisualizationNodeData;
    targetProperty?: string;
  }) => void;

  /** Remove the step at a given path from the underlying Camel entity */
  removeStep: (path?: string) => void;

  /** Returns the NodeInteraction information so the UI can show whether this node can have children and/or siblings */
  getNodeInteraction(data: IVisualizationNodeData): NodeInteraction;

  /** Given a path, retrieve the Node validation status */
  getNodeValidationText(path?: string): string | undefined;

  /** Generates a IVisualizationNode from the underlying Camel entity */
  toVizNode: () => IVisualizationNode;
}

export interface BaseVisualCamelEntityConstructor {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): BaseVisualCamelEntity;
  isApplicable: (entity: unknown) => boolean;
}

/**
 * IVisualizationNode
 *
 * This interface is used to represent a node in the visualization tree.
 * Usually, each IVisualizationNode represent a component in a BaseVisualCamelEntity.
 *
 * This interface shouldn't know anything about the specific BaseVisualCamelEntity.
 */
export interface IVisualizationNode<T extends IVisualizationNodeData = IVisualizationNodeData> {
  id: string;
  data: T;

  /** This property is only set on the root node */
  getBaseEntity(): BaseVisualCamelEntity | undefined;

  /** This method returns the label to be used by the canvas nodes */
  getNodeLabel(labelType?: string): string;

  /** This method returns the tooltip content to be used by the canvas nodes */
  getTooltipContent(): string;

  addBaseEntityStep(definedComponent: DefinedComponent, mode: AddStepMode, targetProperty?: string): void;

  getNodeInteraction(): NodeInteraction;

  getComponentSchema(): VisualComponentSchema | undefined;

  updateModel(value: unknown): void;

  getParentNode(): IVisualizationNode | undefined;

  setParentNode(parentNode?: IVisualizationNode): void;

  getPreviousNode(): IVisualizationNode | undefined;

  setPreviousNode(previousNode: IVisualizationNode): void;

  getNextNode(): IVisualizationNode | undefined;

  setNextNode(node: IVisualizationNode): void;

  getChildren(): IVisualizationNode[] | undefined;

  addChild(child: IVisualizationNode): void;

  removeChild(): void;

  populateLeafNodesIds(ids: string[]): void;

  /** Retrieve the node's validation status, relying into the underlying entity */
  getNodeValidationText(): string | undefined;
}

export interface IVisualizationNodeData {
  icon?: string;
  path?: string;
  entity?: BaseVisualCamelEntity;
  isPlaceholder?: boolean;
  isGroup?: boolean;
  [key: string]: unknown;
}

/**
 * VisualComponentSchema
 *
 * This interface is used to represent a component through
 * the name and the schema of the component.
 */
export interface VisualComponentSchema {
  title: string;
  schema: KaotoSchemaDefinition['schema'];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  definition: any;
}

export const enum AddStepMode {
  /** Used to append a new step before an existing step */
  PrependStep = 'prepend-step',
  /** Used to append a new step after an existing step */
  AppendStep = 'append-step',
  /** Used to replace an existing step */
  ReplaceStep = 'replace-step',
  /**
   * Used to insert a new step in the step property.
   * for instance, for adding steps to the `from` or `doTry` processor
   */
  InsertChildStep = 'insert-step',
  /**
   * Used to insert a new special step in a special step property.
   * for instance, for adding `when` clauses to the `choice.when` property,
   * or to add `doCatch` clauses to the `doTry.doCath` property
   */
  InsertSpecialChildStep = 'insert-special-step',
}

export interface NodeInteraction {
  canHavePreviousStep: boolean;
  canHaveNextStep: boolean;
  canHaveChildren: boolean;
  canHaveSpecialChildren: boolean;
  canReplaceStep: boolean;
  canRemoveStep: boolean;
  canRemoveFlow: boolean;
  canBeDisabled: boolean;
}
