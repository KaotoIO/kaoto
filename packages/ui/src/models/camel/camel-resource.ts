import { BaseVisualCamelEntity } from '../visualization/base-visual-entity';
import { BaseCamelEntity } from './entities';
import { CamelRouteResource } from './camel-route-resource';
import { IntegrationResource } from './integration-resource';
import { KameletResource } from './kamelet-resource';
import { KameletBindingResource } from './kamelet-binding-resource';
import { PipeResource } from './pipe-resource';
import { SourceSchemaType } from './source-schema-type';

export const CAMEL_K_K8S_API_VERSION_V1 = 'camel.apache.org/v1';

export interface CamelResource {
  getVisualEntities(): BaseVisualCamelEntity[];
  getEntities(): BaseCamelEntity[];
  supportsMultipleVisualEntities(): boolean;
  toJSON(): unknown;
  getType(): SourceSchemaType;
  addEntity(entity: BaseCamelEntity): void;
}

/**
 * Creates a CamelResource based on the given {@link type} and {@link json}. If
 * both are not specified, a default empty {@link CamelRouteResource} is created.
 * If only {@link type} is specified, an empty {@link CamelResource} of the given
 * {@link type} is created.
 * @param type
 * @param json
 */
export function createCamelResource(json?: unknown, type?: SourceSchemaType): CamelResource {
  const jsonRecord = json as Record<string, unknown>;
  if (json && typeof json === 'object' && 'kind' in jsonRecord) {
    return doCreateCamelResource(json, jsonRecord['kind'] as SourceSchemaType);
  } else {
    return doCreateCamelResource(json, type || SourceSchemaType.Route);
  }
}

function doCreateCamelResource(json?: unknown, type?: SourceSchemaType): CamelResource {
  switch (type) {
    case SourceSchemaType.Integration:
      return new IntegrationResource(json);
    case SourceSchemaType.Kamelet:
      return new KameletResource(json);
    case SourceSchemaType.KameletBinding:
      return new KameletBindingResource(json);
    case SourceSchemaType.Pipe:
      return new PipeResource(json);
    default:
      return new CamelRouteResource(json);
  }
}
