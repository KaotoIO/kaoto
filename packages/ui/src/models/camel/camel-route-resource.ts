import { RouteDefinition } from '@kaoto-next/camel-catalog/types';
import { ITile, TileFilter } from '../../components/Catalog';
import { isDefined } from '../../utils';
import { CatalogKind } from '../catalog-kind';
import { AddStepMode } from '../visualization/base-visual-entity';
import { CamelRouteVisualEntity, isCamelRoute } from '../visualization/flows';
import { flowTemplateService } from '../visualization/flows/flow-templates-service';
import { CamelRouteVisualEntityData } from '../visualization/flows/support/camel-component-types';
import { BeansEntity, isBeans } from '../visualization/metadata';
import { BeansAwareResource, CamelResource } from './camel-resource';
import { BaseCamelEntity } from './entities';
import { SourceSchemaType } from './source-schema-type';

export class CamelRouteResource implements CamelResource, BeansAwareResource {
  private entities: BaseCamelEntity[] = [];

  constructor(json?: unknown) {
    if (!json) return;
    const rawEntities = Array.isArray(json) ? json : [json];
    this.entities = rawEntities.reduce((acc, rawItem) => {
      const entity = this.getEntity(rawItem);
      if (isDefined(entity) && typeof entity === 'object') {
        acc.push(entity);
      }
      return acc;
    }, [] as BaseCamelEntity[]);
  }

  addNewEntity(_entity?: unknown): string {
    const template = flowTemplateService.getFlowTemplate(this.getType());
    const route = template[0].route as Partial<RouteDefinition>;
    const visualEntity = new CamelRouteVisualEntity(route);
    this.entities.push(visualEntity);

    return visualEntity.id;
  }

  getEntity(rawItem: unknown): BaseCamelEntity | undefined {
    if (!isDefined(rawItem) || Array.isArray(rawItem)) {
      return undefined;
    }
    if (isCamelRoute(rawItem)) {
      return new CamelRouteVisualEntity(rawItem.route);
    } else if (isBeans(rawItem)) {
      return new BeansEntity(rawItem);
    }
    return rawItem as BaseCamelEntity;
  }

  getType(): SourceSchemaType {
    return SourceSchemaType.Route;
  }

  supportsMultipleVisualEntities(): boolean {
    return true;
  }

  getVisualEntities(): CamelRouteVisualEntity[] {
    return this.entities.filter((entity) => entity instanceof CamelRouteVisualEntity) as CamelRouteVisualEntity[];
  }

  getEntities(): BaseCamelEntity[] {
    return this.entities.filter((entity) => !(entity instanceof CamelRouteVisualEntity)) as BaseCamelEntity[];
  }

  toJSON(): unknown {
    return this.entities.map((entity) => entity.toJSON());
  }

  createBeansEntity(): BeansEntity {
    const newBeans = { beans: [] };
    const beansEntity = new BeansEntity(newBeans);
    this.entities.push(beansEntity);
    return beansEntity;
  }

  deleteBeansEntity(entity: BeansEntity): void {
    const index = this.entities.findIndex((e) => e === entity);
    if (index !== -1) {
      this.entities.splice(index, 1);
    }
  }

  removeEntity(id?: string): void {
    const index: number = this.entities.findIndex((e) => e.id === id);

    if (index !== -1) {
      this.entities.splice(index, 1);
    }
    // we don't want to end up with clean entities, so we're adding default one if the list if empty
    if (this.entities.length === 0) {
      this.addNewEntity();
    }
  }

  /** Components Catalog related methods */
  getCompatibleComponents(mode: AddStepMode, visualEntityData: CamelRouteVisualEntityData): TileFilter {
    if (mode === AddStepMode.ReplaceStep && visualEntityData.path === 'from') {
      /**
       * For the `from` step we want to show only components which are not `producerOnly`,
       * as this mean that they can be used only as a consumer.
       */
      return (item: ITile) => {
        return (item.type === CatalogKind.Component && !item.tags.includes('producerOnly')) || (item.type === CatalogKind.Kamelet && item.tags.includes('source'));
      };
    }

    if (mode === AddStepMode.InsertSpecialChildStep) {
      /**
       * specialChildren is a map of processor names and their special children.
       */
      const specialChildren: Record<string, string[]> = {
        choice: ['when', 'otherwise'],
        doTry: ['doCatch', 'doFinally'],
      };

      /**
       * For special child steps, we need to check which type of processor it is, in order to determine
       * what kind of components we want to show.
       */
      return (item: ITile) => {
        if (item.type !== CatalogKind.Processor || specialChildren[visualEntityData.processorName] === undefined) {
          return false;
        }

        return specialChildren[visualEntityData.processorName].includes(item.name);
      };
    }

    /**
     * For the rest, we want to filter out components that are `consumerOnly`,
     * as this mean that they can be used only as a consumer.
     */
    return (item: ITile) => {
      return !item.tags.includes('consumerOnly');
    };
  }
}
