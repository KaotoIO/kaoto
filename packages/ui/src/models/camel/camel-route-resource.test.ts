import { beansJson } from '../../stubs/beans';
import { camelFromJson } from '../../stubs/camel-from';
import { camelRouteJson } from '../../stubs/camel-route';
import { AddStepMode } from '../visualization/base-visual-entity';
import { CamelRouteVisualEntity } from '../visualization/flows/camel-route-visual-entity';
import { BeansEntity } from '../visualization/metadata/beansEntity';
import { createCamelResource } from './camel-resource';
import { CamelRouteResource } from './camel-route-resource';
import { SourceSchemaType } from './source-schema-type';

describe('CamelRouteResource', () => {
  it('should create CamelRouteResource', () => {
    const resource = new CamelRouteResource(camelRouteJson);
    expect(resource.getType()).toEqual(SourceSchemaType.Route);
    expect(resource.getVisualEntities().length).toEqual(1);
    expect(resource.getEntities().length).toEqual(0);
  });

  it('should initialize Camel Route if no args is specified', () => {
    const resource = new CamelRouteResource(undefined);
    expect(resource.getType()).toEqual(SourceSchemaType.Route);
    expect(resource.getEntities()).toEqual([]);
    expect(resource.getVisualEntities()).toEqual([]);
  });

  describe('constructor', () => {
    it.each([
      [camelRouteJson, CamelRouteVisualEntity],
      [camelFromJson, CamelRouteVisualEntity],
      [{ from: { uri: 'direct:foo', steps: [] } }, CamelRouteVisualEntity],
      [{ from: 'direct:foo' }, { from: 'direct:foo' }],
      [{ from: { uri: 'direct:foo' } }, { from: { uri: 'direct:foo' } }],
      [{ beans: [] }, BeansEntity],
      [{}, {}],
      [undefined, undefined],
      [null, undefined],
      [[], undefined],
    ])('should return the appropriate entity for: %s', (json, expected) => {
      const resource = new CamelRouteResource(json);
      const firstEntity = resource.getVisualEntities()[0] ?? resource.getEntities()[0];

      if (typeof expected === 'function') {
        expect(firstEntity).toBeInstanceOf(expected);
      } else {
        expect(firstEntity).toEqual(expected);
      }
    });
  });

  describe('addNewEntity', () => {
    it('should add new entity and return its ID', () => {
      const resource = new CamelRouteResource();
      const id = resource.addNewEntity();

      expect(resource.getVisualEntities()).toHaveLength(1);
      expect(resource.getVisualEntities()[0].id).toEqual(id);
    });
  });

  it('should return the right type', () => {
    const resource = new CamelRouteResource();
    expect(resource.getType()).toEqual(SourceSchemaType.Route);
  });

  it('should allow consumers to have multiple visual entities', () => {
    const resource = new CamelRouteResource();
    expect(resource.supportsMultipleVisualEntities()).toEqual(true);
  });

  it('should return visual entities', () => {
    const resource = new CamelRouteResource(camelRouteJson);
    expect(resource.getVisualEntities()).toHaveLength(1);
    expect(resource.getVisualEntities()[0]).toBeInstanceOf(CamelRouteVisualEntity);
    expect(resource.getEntities()).toHaveLength(0);
  });

  it('should return entities', () => {
    const resource = new CamelRouteResource(beansJson);
    expect(resource.getEntities()).toHaveLength(1);
    expect(resource.getEntities()[0]).toBeInstanceOf(BeansEntity);
    expect(resource.getVisualEntities()).toHaveLength(0);
  });

  describe('toJSON', () => {
    it('should return JSON', () => {
      const resource = new CamelRouteResource(camelRouteJson);
      expect(resource.toJSON()).toMatchSnapshot();
    });

    it.todo('should position the ID at the top of the JSON');
    it.todo('should position the parameters after the ID');
  });

  it('should create beans entity', () => {
    const resource = new CamelRouteResource();
    const beansEntity = resource.createBeansEntity();

    expect(resource.getEntities()).toHaveLength(1);
    expect(resource.getEntities()[0]).toBeInstanceOf(BeansEntity);
    expect(resource.getEntities()[0]).toEqual(beansEntity);
  });

  it('should delete beans entity', () => {
    const resource = new CamelRouteResource();
    const beansEntity = resource.createBeansEntity();

    resource.deleteBeansEntity(beansEntity);

    expect(resource.getEntities()).toHaveLength(0);
  });

  describe('removeEntity', () => {
    it('should not do anything if the ID is not provided', () => {
      const resource = new CamelRouteResource(camelRouteJson);

      resource.removeEntity();

      expect(resource.getVisualEntities()).toHaveLength(1);
    });

    it('should not do anything when providing a non existing ID', () => {
      const resource = new CamelRouteResource(camelRouteJson);

      resource.removeEntity('non-existing-id');

      expect(resource.getVisualEntities()).toHaveLength(1);
    });

    it('should allow to remove an entity', () => {
      const resource = new CamelRouteResource([camelRouteJson, camelFromJson]);
      const camelRouteEntity = resource.getVisualEntities()[0];

      resource.removeEntity(camelRouteEntity.id);

      expect(resource.getVisualEntities()).toHaveLength(1);
    });

    it('should create a new entity after deleting them all', () => {
      const resource = new CamelRouteResource(camelRouteJson);
      const camelRouteEntity = resource.getVisualEntities()[0];

      resource.removeEntity(camelRouteEntity.id);

      expect(resource.getVisualEntities()).toHaveLength(1);
    });
  });

  describe('getCompatibleComponents', () => {
    it('should not provide ProducerOnly components', () => {
      const resource = createCamelResource(camelRouteJson);
      expect(resource.getCompatibleComponents(AddStepMode.ReplaceStep, { path: 'from', label: 'timer' })).toBeDefined();
    });

    it('should not provide consumerOnly components', () => {
      const resource = new CamelRouteResource(camelRouteJson);
      expect(
        resource.getCompatibleComponents(AddStepMode.ReplaceStep, {
          path: 'from.steps.2.to',
          processorName: 'to',
          label: 'timer',
        }),
      ).toBeDefined();
    });

    it('scenario for InsertSpecialChild', () => {
      const resource = createCamelResource(camelRouteJson);
      expect(
        resource.getCompatibleComponents(AddStepMode.InsertSpecialChildStep, { path: 'from', label: 'timer' }),
      ).toBeDefined();
    });

    it('scenario for a new step before an existing step', () => {
      const resource = new CamelRouteResource(camelRouteJson);
      expect(
        resource.getCompatibleComponents(AddStepMode.PrependStep, {
          path: 'from.steps.0.to',
          processorName: 'to',
          label: 'timer',
        }),
      ).toBeDefined();
    });

    it('scenario for a new step after an existing step', () => {
      const resource = new CamelRouteResource(camelRouteJson);
      expect(
        resource.getCompatibleComponents(AddStepMode.AppendStep, {
          path: 'from.steps.1.to',
          processorName: 'to',
          label: 'timer',
        }),
      ).toBeDefined();
    });
  });
});
