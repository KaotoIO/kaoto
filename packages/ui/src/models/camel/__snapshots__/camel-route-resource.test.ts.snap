// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CamelRouteResource toJSON should return JSON 1`] = `
[
  {
    "route": {
      "from": {
        "steps": [
          {
            "set-header": {
              "name": "myChoice",
              "simple": "\${random(2)}",
            },
          },
          {
            "choice": {
              "otherwise": {
                "steps": [
                  {
                    "to": {
                      "uri": "amqp:queue:",
                    },
                  },
                  {
                    "to": {
                      "uri": "amqp:queue:",
                    },
                  },
                  {
                    "log": {
                      "id": "log-2",
                      "message": "We got a \${body}",
                    },
                  },
                ],
              },
              "when": [
                {
                  "simple": "\${header.myChoice} == 1",
                  "steps": [
                    {
                      "log": {
                        "id": "log-1",
                        "message": "We got a one.",
                      },
                    },
                  ],
                },
              ],
            },
          },
          {
            "to": {
              "parameters": {
                "bridgeErrorHandler": true,
              },
              "uri": "direct:my-route",
            },
          },
        ],
        "uri": "timer:tutorial",
      },
      "id": "route-8888",
    },
  },
]
`;
